crates/component-macro/src/lib.rs

```diff
@@ -0,0 +1,447 @@
+    let generics = add_trait_bounds(input.generics);
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let expanded = quote! {
+        unsafe impl #impl_generics wasmtime::component::Lower for #name #ty_generics #where_clause {
+            #[inline]
+            fn lower<T>(
+                &self,
+                store: &mut wasmtime::StoreContextMut<T>,
+                options: &#internal::Options,
+                dst: &mut std::mem::MaybeUninit<Self::Lower>,
+            ) -> #internal::anyhow::Result<()> {
+                #lowers
+                Ok(())
+            }
+
+            #[inline]
+            fn store<T>(
+                &self,
+                memory: &mut #internal::MemoryMut<'_, T>,
+                mut offset: usize
+            ) -> #internal::anyhow::Result<()> {
+                #stores
+                Ok(())
+            }
+        }
+    };
+
+    Ok(expanded)
+}
+
+#[proc_macro_derive(ComponentType, attributes(component))]
+pub fn component_type(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    expand_type(parse_macro_input!(input as DeriveInput))
+        .unwrap_or_else(Error::into_compile_error)
+        .into()
+}
+
+fn expand_type(input: DeriveInput) -> Result<TokenStream> {
+    match find_style(&input)? {
+        Style::Record => expand_record_type(input),
+
+        style => Err(Error::new_spanned(
+            input,
+            format!("todo: expand_type {style:?}"),
+        )),
+    }
+}
+
+fn expand_record_type(input: DeriveInput) -> Result<TokenStream> {
+    let name = &input.ident;
+
+    let body = if let Data::Struct(body) = &input.data {
+        body
+    } else {
+        return Err(Error::new(
+            name.span(),
+            "`record` component types can only be derived for `struct`s",
+        ));
+    };
+
+    let internal = quote!(wasmtime::component::__internal);
+
+    match &body.fields {
+        syn::Fields::Named(fields) => {
+            let mut field_names_and_checks = TokenStream::new();
+            let mut lower_field_declarations = TokenStream::new();
+            let mut sizes = TokenStream::new();
+            let mut alignments = TokenStream::new();
+
+            for field @ syn::Field { ty, ident, .. } in &fields.named {
+                lower_field_declarations
+                    .extend(quote!(#ident: <#ty as wasmtime::component::ComponentType>::Lower,));
+
+                let literal = find_rename(field)?
+                    .unwrap_or_else(|| Literal::string(&ident.as_ref().unwrap().to_string()));
+
+                field_names_and_checks.extend(
+                    quote!((#literal, <#ty as wasmtime::component::ComponentType>::typecheck),),
+                );
+
+                sizes.extend(quote!(#internal::next_field::<#ty>(&mut size);));
+
+                alignments.extend(
+                    quote!(align = align.max(<#ty as wasmtime::component::ComponentType>::align());)
```

Comment:
Random thought: for alignment, you could de-duplicate the field types, e.g. a struct with a bunch of `u32`s has the same alignment as one `u32`. I think it might marginally improve compile times overall. It's certainly not necessary to do though.

Overall, this looks good to me! It's up to Alex on final review though.